contrib/pylightning/lightning/plugin.py

```diff
@@ -0,0 +1,227 @@
+        self.init = None
+
+    def add_method(self, name, func):
+        """Add a plugin method to the dispatch table.
+
+        The function will be expected at call time (see `_dispatch`)
+        and the parameters from the JSON-RPC call will be mapped to
+        the function arguments. In addition to the parameters passed
+        from the JSON-RPC call we add a few that may be useful:
+
+         - `plugin`: gets a reference to this plugin.
+
+         - `request`: gets a reference to the raw request as a
+           dict. This corresponds to the JSON-RPC message that is
+           being dispatched.
+
+        Notice that due to the python binding logic we may be mapping
+        the arguments wrongly if we inject the plugin and/or request
+        in combination with positional binding. To prevent issues the
+        plugin and request argument should always be the last two
+        arguments and have a default on None.
+
+        """
+        if name in self.methods:
+            raise ValueError(
+                "Name {} is already bound to a method.".format(name)
+            )
+
+        # Register the function with the name
+        self.methods[name] = func
+
+    def method(self, method_name, *args, **kwargs):
+        """Decorator to add a plugin method to the dispatch table.
+
+        Internally uses add_method.
+        """
+        def decorator(f):
+            self.add_method(method_name, f)
+            return f
+        return decorator
+
+    def _dispatch(self, request):
+        name = request['method']
+        params = request['params']
+
+        if name not in self.methods:
+            raise ValueError("No method {} found.".format(name))
+
+        args = params.copy() if isinstance(params, list) else []
+        kwargs = params.copy() if isinstance(params, dict) else {}
+
+        func = self.methods[name]
+        sig = inspect.signature(func)
+
+        if 'plugin' in sig.parameters:
+            kwargs['plugin'] = self
+
+        if 'request' in sig.parameters:
+            kwargs['request'] = request
+
+        ba = sig.bind(*args, **kwargs)
+        ba.apply_defaults()
+        return func(*ba.args, **ba.kwargs)
+
+    def notify(self, method, params):
+        payload = {
+            'jsonrpc': '2.0',
+            'method': method,
+            'params': params,
+        }
+        json.dump(payload, self.stdout)
+        self.stdout.write("\n\n")
+        self.stdout.flush()
+
+    def log(self, message, level='info'):
+        self.notify('log', {'level': level, 'message': message})
+
+    def _multi_dispatch(self, msgs):
+        """We received a couple of messages, now try to dispatch them all.
+
+        Returns the last partial message that was not complete yet.
+        """
+        for payload in msgs[:-1]:
+            request = json.loads(payload)
+
+            try:
+                result = {
+                    "jsonrpc": "2.0",
+                    "result": self._dispatch(request),
```

Comment:
It looks to me that this will result in "linear" dispatching, i.e. commands are handled one-at-a-time.

Would it be possible to have this handle dispatching in parallel, i.e. if a command stalls processing, another command can be processed?

Or is some Python limitation exists to prevent this, or it becomes complicated?

(maybe there can be a separate decoration for methods that might block)

Reply:
Good point, hadn't thought about it yet. I wanted to keep away from multi-threading as much as possible, but I'll try to see what I can come up with.

Reply:
I opened issue #2167 to track progress on this. Would love to hear feedback for the initial idea, which is future-inspired, but since the python futures are tightly bound to multi-threading I'm thinking of implementing my own `AsyncResult` instead.

Reply:
This seems sensible.